storage/country_tree.hpp

```diff
@@ -6,205 +6,235 @@
+      for (auto & child : m_children)
+        child->ForEachInSubtree(f);
+    }
 
-  CountryTree<T> const & Parent() const
-  {
-    CHECK(HasParent(), ());
-    return *m_parent;
-  }
+    template <class TFunctor>
+    void ForEachInSubtree(TFunctor && f) const
+    {
+      f(*this);
+      for (auto const & child : m_children)
+        child->ForEachInSubtree(f);
+    }
 
-  CountryTree<T> const & Child(size_t index) const
-  {
-    ASSERT_LESS(index, m_children.size(), ());
-    return *m_children[index];
-  }
+    template <class TFunctor>
+    void ForEachAncestorExceptForTheRoot(TFunctor && f)
+    {
+      if (m_parent == nullptr || m_parent->m_parent == nullptr)
+        return;
+      f(*m_parent);
+      m_parent->ForEachAncestorExceptForTheRoot(f);
+    }
 
-  size_t ChildrenCount() const
-  {
-    return m_children.size();
-  }
+    template <class TFunctor>
+    void ForEachAncestorExceptForTheRoot(TFunctor && f) const
+    {
+      if (m_parent == nullptr || m_parent->m_parent == nullptr)
+        return;
+      f(*m_parent);
+      m_parent->ForEachAncestorExceptForTheRoot(f);
+    }
+  };
 
-  /// \brief Calls functor f for each first generation descendant of the node.
-  template <class TFunctor>
-  void ForEachChild(TFunctor && f)
-  {
-    for (auto & child : m_children)
-      f(*child);
-  }
+private:
+  using TCountryTreeHashTable = unordered_multimap<TKey, Node *>;
 
-  template <class TFunctor>
-  void ForEachChild(TFunctor && f) const
+public:
+  CountryTree(TValue const & value = TValue(), Node * parent = nullptr)
+    : m_countryTree(make_unique<Node>(value, parent))
   {
-    for (auto const & child : m_children)
-      f(*child);
   }
 
-  /// \brief Calls functor f for all nodes (add descendant) in the tree.
-  template <class TFunctor>
-  void ForEachDescendant(TFunctor && f)
+  Node const & GetRoot() const { return *m_countryTree; }
+  Node & GetRoot() { return *m_countryTree; }
+
+  /// @return reference is valid only up to the next tree structure modification
+  TValue & AddAtDepth(int level, TValue const & value)
   {
-    for (auto & child : m_children)
-    {
-      f(*child);
-      child->ForEachDescendant(f);
-    }
+    Node * const added = m_countryTree->AddAtDepth(level, value);
+    ASSERT(added, ());
+    m_countryTreeHashTable.insert(make_pair(value.Name(), added));
+    return added->Value();
   }
 
-  template <class TFunctor>
-  void ForEachDescendant(TFunctor && f) const
+  /// Deletes all children and makes tree empty
+  void Clear() { m_countryTree->Clear(); }
+
+  /// \brief Checks all nodes in tree to find an equal one. If there're several equal nodes
+  /// returns the first found.
+  /// \returns a poiter item in the tree if found and nullptr otherwise.
+  void Find(TValue const & value, vector<Node const *> & found) const
   {
-    for (auto const & child: m_children)
+    found.clear();
+
+    if (IsEqual(value, m_countryTree->Value()))
+      found.push_back(m_countryTree.get());
```

Comment:
К сожалению, наше дерево сейчас построено так, что его корень обрабатывается отдельным образом. Причины тому, отчасти исторические (см. CountryTree::Clear и пр.) от части от того, что я до сих пор не переделал, а теперь разные такие if-фы растут, как снежный ком.

> Не совсем понятно что мы тут отдельно проверяем и для чего? Паз у нас и так уже хранится hash table.

Все кроме корня. Мы заполняем хэш таблицу в CountryTree::Add(), а корень иначе.

Быстро это не переделать см.: LoadCountries, CountryTree::Clear, SetCountriesContainerAttrs и т.д.

Думаю, это тема отдельного PR. А пока, я оставил вот такое TODO: 

  /// @TODO(bykoianko) The root of the tree currently is processed in a special way.
  /// It's never deleted. Because of it it's necessary to work with the root in a special way.
  /// See SetCountriesContainerAttrs method for example. And CountryTree::Clear() method
  /// does not delete the root. It should be fixed.
  unique_ptr<Node> m_countryTree;